torch/nn/modules/activation.py

```diff
@@ -669,6 +673,225 @@ def extra_repr(self):
         return str(self.lambd)
 
 
+@weak_module
+class MultiheadAttention(Module):
+    r"""Allows the model to jointly attend to information
+    from different representation subspaces.
+    See reference: Attention Is All You Need
+
+    .. math::
+        \text{MultiHead}(Q, K, V) = \text{Concat}(head_1,\dots,head_h)W^O
+        \text{where} head_i = \text{Attention}(QW_i^Q, KW_i^K, VW_i^V)
+
+    Args:
+        embed_dim: total dimension of the model
+        num_heads: parallel attention layers, or heads
+
+    Examples::
+
+        >>> multihead_attn = nn.MultiheadAttention(embed_dim, num_heads)
+        >>> output = multihead_attn(query, key, value)
+    """
+
+    def __init__(self, embed_dim, num_heads, dropout=0., bias=True, add_bias_kv=False, add_zero_attn=False):
+        super(MultiheadAttention, self).__init__()
+        self.embed_dim = embed_dim
+        self.num_heads = num_heads
+        self.dropout = dropout
+        self.head_dim = embed_dim // num_heads
+        assert self.head_dim * num_heads == self.embed_dim, "embed_dim must be divisible by num_heads"
+        self.scaling = self.head_dim ** -0.5
+
+        self.in_proj_weight = Parameter(torch.Tensor(3 * embed_dim, embed_dim))
```

Comment:
Fairseq softmax employs this perf optimization that is carried over here, where 3 matrix multiplies are combined into 1, with a bigger weight (similar to how this is done e.g. for LSTM). Mathematically, the operations performed are exactly the same, however, initialization could be different if it is based on the first dimension of the parameter matrix. IIRC, tf t2t transformer has 3 weight matrices here instead of one, with correspondingly different initialization. I'm not asking to change anything for this PR, I just want us to be aware of cross-framework initialization differences, and of potential changes that might happen here if we decide to go to 3 matrices instead of 1 (believe it or not, there can be perf reasons to go back to 3 matrix multiplies).

Reply:
This might be important to decide at this point right? Presumably this will be hard to change in the future, unless we create an option to switch between the two.

Reply:
If we want core multihead attention to replicate the behavior of fairseq, then our hands are tied already. I'd defer to @myleott for a question of what initialization behavior we want, and what should be the options.

Reply:
@ngimel makes a great point. We can decouple these, by initializing the weights of each portion separately.

Lemme run a quick test to see how this affects NMT model quality. I’ll share findings back here ~tomorrow.

Reply:
Hi @myleott. I just want to reach out and see if the decision has been made.

Reply:
It actually does seem to make a difference, with the fairseq version producing better results by ~0.9 BLEU on WMT'16 En-De. It's possible that this could be corrected by adjusting the learning rate, although I didn't try this yet.

It's up to you how to proceed. We'll have fairseq switch to this implementation in any case, and use a custom weight init if needed.

Reply:
Thanks for the update @myleott . I will change to three individual matrices for the weights and initiate them separately.

Reply:
I'm not sure you should actually make it three separate matrices, since that will be slower. You can instead initialize them as if there were separate though:
```
nn.init.xavier_uniform_(self.in_proj_weight[:embed_dim, :])
nn.init.xavier_uniform_(self.in_proj_weight[embed_dim:2*embed_dim, :])
nn.init.xavier_uniform_(self.in_proj_weight[2*embed_dim:, :])
```

Reply:
I looked into this a bit more and it's actually a much smaller difference, only ~0.3 BLEU worse using the split init vs. combined init.

Reply:
I will just keep it as the split init. Thanks for investigating this.